src/sagemaker_xgboost_container/algorithm_mode/callback.py

```diff
@@ -0,0 +1,205 @@
+
+
+class SaveCheckpoint:
+    """Create a callback that saves checkpoints to disk.
+
+    This class represents the hook which is meant to be used a callback
+    function in XGBoost.
+
+    Example
+    -------
+    >>> save_checkpoint = SaveCheckpoint("/opt/ml/checkpoints")
+    >>> xgboost.train(prams, dtrain, callbacks=[save_checkpoint])
+
+    At the end of each iteration, we save each checkpoint to different files
+    parametrized by the iteration number, e.g., checkpoint.01, checkpoint.02,
+    etc. However, this could result in a large number of files to save locally
+    (and on S3 if used in SageMaker). To save disk space and reduce the amount
+    of data required to be downloaded on resumption, we retain only the N
+    (default 5, spcified by 'max_to_keep') most recent checkpoints.
+
+    To delete stale checkpoints, we use a producer-consumer pattern: we start a
+    daemon thread in the background and maintain a queue of files to delete.
+    There may be a lock on the file if SageMaker is uploading the file; in that
+    case, the file is put back on the queue and we try again later. When
+    training is complete, we put SENTINEL on the queue, and when we see the
+    SENTINEL, we clean up and exit the thread.
+    """
+
+    SENTINEL = None
+
+    def __init__(
+            self, checkpoint_dir: str, max_to_keep: int = 5, start_iteration: int = 0,
+            num_round: Optional[int] = None
+            ) -> None:
+        """
+        Parameters
+        ----------
+        checkpoint_dir: indicates the path to the directory where checkpoints
+            will be saved.  Defaults to /opt/ml/checkpoints on SageMaker.
+        max_to_keep: indicates the maximum number of recent checkpoint files to
+            keep.  As new files are created, older files are deleted.  Defaults
+            to 5 (that is, the 5 most recent checkpoint files are kept.)
+        start_iteration: indicates the round at which the current training
+            started. If xgb_model was loaded from a previous checkpoint, this
+            will be greater than 0 (that is, if the previous training ended
+            after round 19, start_iteration will be 20).
+        num_round: (optional) indicates the number of boosting rounds.
+        """
+        self.checkpoint_dir = checkpoint_dir
+        self.max_to_keep = max_to_keep
+        self.start_iteration = start_iteration
+        self.num_round = num_round
+
+        self.step = start_iteration
+        self.previous_checkpoints = [
+            os.path.join(self.checkpoint_dir, f)
+            for f in os.listdir(self.checkpoint_dir)]
+
+        self.thread = None
+        self.delete_queue = queue.Queue()
+
+        self.start()
+
+    def __call__(self, env: CallbackEnv) -> Callable[[CallbackEnv], None]:
+
+        if not os.path.exists(self.checkpoint_dir):
+            os.makedirs(self.checkpoint_dir)
+
+        return self.callback(env)
+
+    def fmt_path(self, i: int) -> str:
```

Comment:
What is i? Why are we zero_padding the file name?

Reply:
Also, if we are keeping this method, can you just rename it to format_path? You don't gain anything by removing the couple of letters, and you lose readability later.

Reply:
Will change the variable and function names to be more descriptive.

On zero padding, I used the same format that tensorboard uses to save tfevent files (I just happened to be working on something else that uses tensorboard at the time). I think the rationale is by zero padding, it becomes string-sortable.

Reply:
If that is the case, can you describe the rational for how much zero padding you want to do and document? Sortability is nice, but if you have a limited number of zero padding it affects scaleability. In general I would rather opt for scaleability rather than sortability; also why would users have to sort the checkpoints?

Reply:
By scaleability, do you mean we might run out of zeros, like `.999999999999` -> `.000000000000`. Choosing 12 digits was somewhat arbitrary, and I don't expect anyone to train for trillion rounds.

Sortability is useful in `train_utils.load_checkpoint()` where I sort the checkpoints and try to load the latest checkpoint, but that can easily be changed if we remove zero-padding. I think you have a point; from the user's perspective, zero-padding is unnecessary.

It's a valid point, so I could get rid of zero-padding and modify the sorting method in `load_checkpoint()`. Do you think that's better?

Reply:
I agree with the trillion rounds, but no need to add limits that aren't necessary.

I think we should remove the zero-padding if there is no customer usecase for it.